File: _legacy_rn/lib/auth/socialAuth.ts
import { makeRedirectUri } from 'expo-auth-session';
import * as Crypto from 'expo-crypto';
import * as WebBrowser from 'expo-web-browser';
import { Platform } from 'react-native';

import { supabase } from '../supabase/client';

WebBrowser.maybeCompleteAuthSession();

// Apple Sign-In — iOS only, native sheet. No deep link needed.
//
// A random nonce is generated per attempt and its SHA-256 hash is sent to
// Apple in the authorization request; Apple embeds that hash in the signed
// identity token it returns. The *raw* nonce is then passed to Supabase,
// which independently hashes it and checks it against the token's `nonce`
// claim before trusting the token. Without this, a captured/replayed Apple
// identity token (e.g. exfiltrated from another app or a MITM'd request)
// could be replayed against signInWithIdToken to mint an Anchor session for
// an identity the caller doesn't actually control at that moment — the
// nonce binds the token to *this specific* sign-in attempt.
export async function signInWithApple() {
  // Lazy-load so the module is safe to import on Android (where it's absent).
  const AppleAuth = require('expo-apple-authentication');

  const available = await AppleAuth.isAvailableAsync();
  if (!available) throw new Error('Apple Sign-In is not available on this device.');

  const rawNonce = Crypto.randomUUID();
  const hashedNonce = await Crypto.digestStringAsync(Crypto.CryptoDigestAlgorithm.SHA256, rawNonce);

  const credential = await AppleAuth.signInAsync({
    requestedScopes: [
      AppleAuth.AppleAuthenticationScope.FULL_NAME,
      AppleAuth.AppleAuthenticationScope.EMAIL,
    ],
    nonce: hashedNonce,
  });

  const { data, error } = await supabase.auth.signInWithIdToken({
    provider: 'apple',
    token: credential.identityToken,
    nonce: rawNonce,
  });

  if (error) throw error;
  return data;
}

// Google Sign-In — opens an in-app browser, redirects back via deep link.
// Requires EXPO_PUBLIC_SUPABASE_URL to be configured (already is).
// The Supabase project must have Google OAuth enabled and the redirect URI
// anchor://auth/callback listed as allowed.
export async function signInWithGoogle() {
  const redirectUri = makeRedirectUri({ scheme: 'anchor', path: 'auth/callback' });

  const { data, error } = await supabase.auth.signInWithOAuth({
    provider: 'google',
    options: {
      redirectTo: redirectUri,
      skipBrowserRedirect: true,
    },
  });

  if (error) throw error;
  if (!data.url) throw new Error('No OAuth URL returned from Supabase.');

  const result = await WebBrowser.openAuthSessionAsync(data.url, redirectUri);

  if (result.type === 'success') {
    const { error: sessionError } = await supabase.auth.exchangeCodeForSession(result.url);
    if (sessionError) throw sessionError;
  }
  // result.type === 'cancel' or 'dismiss' — user cancelled, do nothing
}

// True on iOS devices that support Apple Sign-In (iOS 13+).
export function isAppleSignInAvailable(): boolean {
  return Platform.OS === 'ios';
}

File: lib/core/supabase/social_auth.dart
import 'dart:convert';
import 'dart:io' show Platform;
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'package:sign_in_with_apple/sign_in_with_apple.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import 'supabase_setup.dart';

/// True on iOS, where the native Apple sheet is available.
bool get isAppleSignInAvailable => Platform.isIOS;

String _randomNonce([int length = 32]) {
  const chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVXYZabcdefghijklmnopqrstuvwxyz-._';
  final rnd = Random.secure();
  return List.generate(length, (_) => chars[rnd.nextInt(chars.length)]).join();
}

/// Apple Sign-In (iOS): a random nonce is generated per attempt and its
/// SHA-256 hash sent to Apple, which embeds it in the signed identity token.
/// The *raw* nonce goes to Supabase, which hashes it independently and checks
/// it against the token's `nonce` claim - binding the token to this specific
/// sign-in attempt so a captured token can't be replayed.
Future<void> signInWithApple() async {
  final rawNonce = _randomNonce();
  final hashed = sha256.convert(utf8.encode(rawNonce)).toString();

  final credential = await SignInWithApple.getAppleIDCredential(
    scopes: [AppleIDAuthorizationScopes.email, AppleIDAuthorizationScopes.fullName],
    nonce: hashed,
  );
  final idToken = credential.identityToken;
  if (idToken == null) throw const AuthException('Apple did not return an identity token.');

  await supabase.auth.signInWithIdToken(
    provider: OAuthProvider.apple,
    idToken: idToken,
    nonce: rawNonce,
  );
}

bool _googleInitialized = false;

/// Google Sign-In uses the native account picker and exchanges the resulting
/// ID token with Supabase (no browser/deep-link round trip). Requires
/// GOOGLE_WEB_CLIENT_ID in .env and Google enabled in the Supabase project.
Future<void> signInWithGoogle() async {
  final serverClientId = dotenv.maybeGet('GOOGLE_WEB_CLIENT_ID');
  if (serverClientId == null || serverClientId.isEmpty) {
    throw const AuthException('Google Sign-In is not configured (GOOGLE_WEB_CLIENT_ID).');
  }

  final google = GoogleSignIn.instance;
  if (!_googleInitialized) {
    await google.initialize(serverClientId: serverClientId);
    _googleInitialized = true;
  }

  try {
    final account = await google.authenticate();
    final idToken = account.authentication.idToken;
    if (idToken == null) throw const AuthException('Google did not return an ID token.');
    await supabase.auth.signInWithIdToken(provider: OAuthProvider.google, idToken: idToken);
  } on GoogleSignInException catch (e) {
    // User cancelled: do nothing.
    if (e.code == GoogleSignInExceptionCode.canceled) return;
    rethrow;
  }
}
